---
title: "Bismark Quality Checks"
author: "Sarah Mueller"
date: "2025 - final"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
#Libraries
library(tidyverse)
library(ggplot2)
library(viridis)
library(wesanderson)
library(RColorBrewer)
library(dplyr)
library(reshape2)
library(viridis)
library(gridExtra)
library(ggsci)
library(ggpubr)

setwd("/dss/dsslegfs01/pr53da/pr53da-dss-0034/projects/2021_SwallowRRBS/")
```

## Methylation Alignment 

Look at the percentage of mapped readsd from the multiqc report on the bismark methylation extraction files. 

Note: The working directory should be set to 2021_SwallowRRBS, everything runs from there.

```{r}
##map percent aligned from multiqc
df.multiqc <- read.table(file = "./0_Pre-processing/0_quality/2_bismark.map.extraction/multiqc_general_stats.txt", header=T)
colnames(df.multiqc)<- c("Sample_ID","p_cpg","p_chg", "p_chh", "total_reads", "aligned_reads", "percent_aligned")

df.multiqc <- df.multiqc %>%
  separate(Sample_ID, into= c("Sample_ID", "SRR", "read", "id"), sep =  "__")

ggplot(df.multiqc,aes(x=Sample_ID,y=percent_aligned))+
  geom_bar(stat="identity",position = "dodge")+
  theme_classic(base_size=16)+
  ylim(0,100)+
  ggtitle("% Aligned Reads using Bismark")+
  theme(axis.text.x=element_text(angle=90),axis.title.x=element_blank())
```

## Methylation Filtering

After using this script (./0_Pre-processing/2_bismark_filter/1_bismark.filter.merged.sh), I want to look at how man CpG loci were filtered out in each filtering step (coverage, SNP, repeats, etc.) 

```{r}
df <- read.delim(file = "0_Pre-processing/2_bismark_filter/1_output_files/a.COUNTS.clean.txt", sep= " ", header=T)

#need to get data into a workable format
df.melt <- melt(df)

df.summary <- df.melt %>% 
  group_by(variable) %>%
  summarize(mean=mean(value),sd=sd(value))

#make color palette
pal <- c( wes_palette("Darjeeling1")[1],
          wes_palette("Cavalcanti1")[2],
          wes_palette("Darjeeling1")[5],
          wes_palette("Darjeeling2")[2],
          wes_palette("FantasticFox1")[2],
          wes_palette("FantasticFox1")[5],
          "grey33")

##make histogram
ggplot(df.melt,aes(x=SAMPLE_ID, fill=variable, y=value))+
  geom_bar(stat="identity",position="dodge")+
  theme_classic(base_size=18)+
  scale_fill_manual(values=pal)+
  ylab("Number of CpGs")+
  theme(axis.text.x=(element_text(angle=90,hjust=1)), axis.title.x=element_blank())


#Specifically, how many SNPs were filtered out
df <- df %>%
  mutate(snps_filtered = CUT_SITE_FILTER - SNP_FILTER)

summary(df$snps_filtered)

##Total Filtered
df <- df %>%
  mutate(loci_filtered = RAW - REPEAT_FILTER)

summary(df$loci_filtered)
```

## M-bias plotting

I put the M-bias output into the bismark.extraction quality folder within the fastqc

```{r}
#plotting m-bias plots, Change the input file based on R1 or R2, or subset the file if you want to look at specific Sample Ids)
bias <- read.table('Methylation_Bias-Input_R2_subset1.txt',header=T)

a <- ggplot(bias,aes(x=Position,col=Sample))+
  geom_line(aes(y=PercentMethylated),stat="identity",show.legend=F,size=0.5)+ylab('Percent Methylation')+
  theme_classic(base_size=16)+
  scale_color_viridis(discrete=T, option = 'turbo')+
  coord_cartesian(ylim=c(0,100))+
  ggtitle('Methylation Along Read Length R2')

a

b <- ggplot(bias,aes(x=Position,col=Sample,y=log(Coverage)))+
  geom_line(stat="identity",show.legend=F,size=0.5)+ylab('log(Coverage)')+
  theme_classic(base_size=16)+
  scale_color_viridis(discrete=T,option='turbo')+
  ggtitle('Coverage Along Read Length')
b

png('M-Bias.png',units='in',height=8,width=12,res = 300)
ggarrange(a,b,
          ncol=2)
dev.off()
```

##5.7 Sex identification check

```{r}
#Create genome 
##change dependign on which chromosomes you want to be represented
bs.genome <- toGRanges(data.frame(chr=c("chr01","chrZ"), start=c(1,1), end=c(119023420,90132486)))

cov <- read.table("Sex_ID.raw.txt",header=FALSE)
head(cov)

cov_GR <- makeGRangesFromDataFrame(df=cov,ignore.strand=TRUE,seqinfo = NULL, seqnames.field = "V2",start.field = "V3", end.field = "V3",starts.in.df.are.0based=FALSE, keep.extra.columns=TRUE)
cov_GR

samples <- as.vector(unique(cov_GR$V1))
samples

total.tracks <- 28

#Plot
pp <- getDefaultPlotParams(plot.type=4)
pp$leftmargin <- 0.09
kp <- plotKaryotype(plot.type=4, genome = bs.genome, main = "chr01 & chrZ coverage differences",plot.params = pp,labels.plotter=NULL)
kpAddChromosomeNames(kp,srt=90, cex=1)


for (i in 1:10) {
  name <- samples[i]
  sub <- cov_GR[grepl(name,cov_GR$V1), ]
  at <- autotrack(current.track = i, total.tracks = 10)
  kpPoints(kp,data=sub,y=sub$V5,cex=0.75,col="black",pch=16,r0=at$r0,r1=at$r1,ymin=0,ymax=200)
  kpAxis(kp,ymin=0,ymax=200,r0=at$r0,r1=at$r1)
}

#histogram of coverage between two chromsomes, or more depending on how you plot
#mean and std of chromosomes

sexes <- cov %>% 
  group_by(V1,V2) %>%
  summarise(mean = mean(V5),n=n())

sexes <- arrange(sexes, V6)
head(sexes)

write.table(sexes, file="sex.id.sorted.txt")

sexes.f <- read.table("sex.id.sorted.f.txt", header=T)
sexes.m <- read.table("sex.id.sorted.m.txt", header=T)
sexes.u <- read.table("sex.id.sorted.u.txt", header=T)

ggplot(sexes,aes(x=V1,y=mean,fill=V2))+
  geom_bar(stat="identity",position = "dodge")+
  theme_classic(base_size=16)+
  ggtitle("Sex ID from Bismark Coverage Files")+
  theme(axis.text.x=element_text(angle=90),axis.title.x=element_blank())

```

##Coverage along the genome

```{r}
#Create genome 
#can change depending on the chromsomes examined
bs.genome <- toGRanges(data.frame(chr=c("chr01"), start=c(1), end=c(119023420)))

cov <- read.table("chr01.txt",header=FALSE)
head(cov)

cov_GR <- makeGRangesFromDataFrame(df=cov,ignore.strand=TRUE,seqinfo = NULL, seqnames.field = "V2",start.field = "V3", end.field = "V3",starts.in.df.are.0based=FALSE, keep.extra.columns=TRUE)
cov_GR

samples <- as.vector(unique(cov_GR$V1))
samples

total.tracks <- 28

#Plot
pp <- getDefaultPlotParams(plot.type=4)
pp$leftmargin <- 0.09
kp <- plotKaryotype(plot.type=4, genome = bs.genome, main = "chr01 coverage",plot.params = pp,labels.plotter=NULL)
kpAddChromosomeNames(kp,srt=90, cex=1)


for (i in 1:10) {
  name <- samples[i]
  sub <- cov_GR[grepl(name,cov_GR$V1), ]
  at <- autotrack(current.track = i, total.tracks = 10)
  kpPoints(kp,data=sub,y=sub$V5,cex=0.75,col="black",pch=16,r0=at$r0,r1=at$r1,ymin=0,ymax=200)
  kpAxis(kp,ymin=0,ymax=200,r0=at$r0,r1=at$r1)
}

#histogram of coverage between two chromsomes, or more depending on how you plot
#mean and std of chromosomes

chr01 <- cov %>% 
  group_by(V1,V2) %>%
  summarise(mean = mean(V5),n=n())


ggplot(chr01,aes(x=factor(V1),y=mean,fill=V2))+
  geom_bar(stat="identity",position = "dodge")+
  theme_classic(base_size=16)+
  ggtitle("Average Coverage across Chromosome 01 per Individual")+
  theme(axis.text.x=element_text(angle=90),axis.title.x=element_blank(), legend.position="none")


coverage <- read.table("coverage.txt", header=T)

ggplot(coverage,aes(x=factor(Sample.Id),y=coverage,fill=wes_palette("Darjeeling1")[1]))+
  geom_bar(stat="identity",position = "dodge")+
  theme_classic(base_size=16)+
  ggtitle("Average coverage across all chromsomes per technical replicate")+
  theme(axis.text.x=element_text(angle=90),axis.title.x=element_blank(), legend.position="none")

```
